### PR TITLE
ENH: optimize._chandrupatla: add array API support

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-7242

#### What does this implement/fix?
This adds array API support to `scipy.optimize._chandrupatla` and paves the way toward adding array API support to the other elementwise iterative methods (e.g. `_differentiate`, `_tanhsinh`, `_nsum`, `_chandrupatla_minimize`, and the bracket finders). I'll propose making the rootfinder, minimizer, and bracket finders public shortly.

#### Additional information
The performance of this function compared to the existing bracketing rootfinders is worst when function evaluations are inexpensive, so I compared the performance between
- `brentq`
- `_chandrupatla` (NumPy)
- `_chandrupatla` (PyTorch, CPU)
- `_chandrupatla` (CuPy)

when finding the root of `xp.cos(x) - p` for many values of `p`. The bracket is always $[0, \pi]$. `brentq`'s default absolute x-tolerance is `2e-12`, so I set `_chandrupatla` to `1e-12` to be more than fair, and I confirmed that both solvers meet their respective tolerance. `_chandrupatla` finds the roots in a single vectorized call whereas `brentq` loops with a list comprehension.

image

<details>

```
# import time
# import numpy as np
# import cupy as cp
# import torch
# import matplotlib.pyplot as plt
#
# rng = np.random.default_rng(1638083107694713882823079058616272161)
# from scipy import optimize
# from scipy.optimize._chandrupatla import _chandrupatla
#
# xp = torch
# n = 10
# n = int(n)
# a = xp.asarray(0)
# b = xp.asarray(np.pi)
# ps = xp.asarray(rng.random(size=n) * 2 - 1, dtype=xp.float64)
#
# def f(x, p):
#     return xp.cos(x) - p
#
# tic = time.perf_counter()
# res = _chandrupatla(f, a, b, args=(ps,), xatol=1e-12)
# toc = time.perf_counter()

import time
import numpy as np
import cupy as cp
import torch
import matplotlib.pyplot as plt

rng = np.random.default_rng(1638083107694713882823079058616272161)
from scipy import optimize
from scipy.optimize._chandrupatla import _chandrupatla

ns = np.logspace(0, 6, 30)

times = {np: [], cp: [], torch: [],  'brentq':[]}

for xp in [np, cp, torch, 'brentq']:
    brentq = False
    if xp == 'brentq':
        xp = np
        brentq = True

    for n in ns:
        n = int(n)
        a = xp.asarray(0)
        b = xp.asarray(np.pi)
        ps = xp.asarray(rng.random(size=n) * 2 - 1, dtype=xp.float64)

        def f(x, p):
            return xp.cos(x) - p

        if brentq:
            tic = time.perf_counter()
            ref = [optimize.brentq(f, a, b, args=(p,)) for p in ps]
            toc = time.perf_counter()
            times['brentq'].append(toc - tic)
            np.testing.assert_allclose(ref, np.arccos(ps), atol=2e-12)
            continue

        tic = time.perf_counter()
        res = _chandrupatla(f, a, b, args=(ps,), xatol=1e-12)
        toc = time.perf_counter()
        times[xp].append(toc - tic)
        np.testing.assert_allclose(cp.asnumpy(res.x), np.arccos(cp.asnumpy(ps)), atol=1e-12)

plt.loglog(ns, times[np], label='np')
plt.loglog(ns, times[cp], label='cp')
plt.loglog(ns, times[torch], label='torch')
plt.loglog(ns, times['brentq'], label='brentq')
plt.xlabel('number of roots')
plt.ylabel('execution time (s)')
plt.title('Root of `xp.cos(x) - p`')
plt.legend()
plt.show()
```

</details>

The function has a lot of overhead due to bells and whistles (e.g. input validation with nice error messages, rich result object, callback function support, etc.). But for solving a lot of equations, the overhead of function calls eventually becomes problematic for `brentq`.

In this case, you could probably get better performance with `cython_optimize`, but for more expensive functions with overhead, like finding the `argus(1)` distribution ppf (https://github.com/scipy/scipy/pull/17719#issuecomment-1547778146), the advantage is much more pronounced.

image

<details>

```python3
import time
import numpy as np
import cupy as cp
import torch
import matplotlib.pyplot as plt

rng = np.random.default_rng(1638083107694713882823079058616272161)
from scipy import optimize, stats
from scipy.optimize._chandrupatla import _chandrupatla

ns = np.logspace(0, 4, 30)

times = {np: [], cp: [], torch: [], 'brentq':[]}

for xp in [np, 'brentq']:
    brentq = False
    if xp == 'brentq':
        xp = np
        brentq = True

    for n in ns:
        n = int(n)
        a = 0.001
        b = 0.999
        ps = np.linspace(0.005, 0.995, n)

        dist = stats.argus(1)
        def f(x, p):
            return dist.cdf(x) - p

        if brentq:
            tic = time.perf_counter()
            ref = [optimize.brentq(f, a, b, args=(p,)) for p in ps]
            toc = time.perf_counter()
            times['brentq'].append(toc - tic)
            continue

        tic = time.perf_counter()
        res = _chandrupatla(f, a, b, args=(ps,), xatol=1e-12)
        toc = time.perf_counter()
        times[xp].append(toc - tic)

plt.loglog(ns, times[np], label='np')
# plt.loglog(ns, times[cp], label='cp')
plt.loglog(ns, times['brentq'], label='brentq')
plt.xlabel('number of probabilities')
plt.ylabel('execution time (s)')
plt.title('Root of `argus(1).cdf(x) - p`')
plt.legend()
plt.show()
```

</details>

`newton` is definitely faster than this function, but it should be - the algorithm converges faster. (The advantage of bracketing methods is that convergence is guaranteed if the bracket is valid.) We can add that as another method with the same framework.

~~The tests are quite strict; I'm ironing out a few failures with alternative backends.~~ Done if CI looks good.

The function currently uses fancy indexing assignment. I can investigate working around that for strict array API support later.

### CODE REVIEW

File: scipy/optimize/_chandrupatla.py
Comment (tupui) [inline scipy/optimize/_chandrupatla.py:1]: And I guess this might go into `optimize.elementwise`

**Review by tupui — COMMENTED** [review]
Nice, I just have a few questions, the rest is pretty straightforward and LGTM 👍

File: scipy/_lib/_array_api.py [hunk 1]
```diff
@@ -428,3 +431,19 @@ def xp_copysign(x1, x2, xp=None):
...
+
+# partial substitute for xp.sign, which does not cover the NaN special case
+# that I need. (https://github.com/data-apis/array-api-compat/issues/136)
+def xp_sign(x, xp=None):
+    xp = array_namespace(x) if xp is None else xp
+
+    if is_numpy(xp):
+        return xp.sign(x)
+
+    sign = xp.full_like(x, xp.nan)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/_lib/_array_api.py:444]: There should be NaNs if none of the comparisons are true. See data-apis/array-api-compat#136

File: scipy/_lib/_elementwise_iterative_method.py [hunk 2]
```diff
@@ -89,38 +92,38 @@ def _initialize(func, xs, args, complex_ok=False, preserve_shape=None):
...
 
     # To ensure that we can do indexing, we'll work with at least 1d arrays,
     # but remember the appropriate shape of the output.
-    xs = [x.ravel() for x in xs]
-    fs = [f.ravel() for f in fs]
-    args = [arg.flatten() for arg in args]
-    return func, xs, fs, args, shape, xfat
+    xs = [xp.reshape(x, (-1,)) for x in xs]
+    fs = [xp.reshape(f, (-1,)) for f in fs]
+    args = [xp.reshape(xp.asarray(arg, copy=True), (-1,)) for arg in args]
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/_lib/_elementwise_iterative_method.py:120]: I think this was just general protection for all elementwise iterative methods to avoid the possibility of them modifying user-provided arrays.

File: scipy/optimize/_chandrupatla.py [hunk 3]
```diff
@@ -1,7 +1,16 @@
...
 from ._zeros_py import _rtol
 import scipy._lib._elementwise_iterative_method as eim
 from scipy._lib._util import _RichResult
+from scipy._lib._array_api import xp_clip, xp_minimum, xp_sign
+
+# TODO:
+# - debug test_maxiter_callback
+# - debug torch failure in test_special_cases - `float64` is fine?
+# - (maybe?) don't use fancy indexing assignment
+# - figure out how to replace the new `try`/`except`s
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/optimize/_chandrupatla.py:12]: It's not strictly correct, I don't think. The try/excepts are used to identify compressible arrays - the ones that should contain only "active" elements. But I can envision a case in which there is an attribute that is stored as a 1D array that *happens* to have the same size as the mask and gets compressed even though it's not supposed to. I need to investigate more closely. It would probably be better for the author using the framework to specify the attribute names of compressible arrays.

File: scipy/optimize/_chandrupatla.py [hunk 4]
```diff
@@ -162,37 +173,38 @@ def check_termination(work):
...
+        i = xp.abs(work.fmin) <= work.fatol + work.frtol
         work.status[i] = eim._ECONVERGED
         stop[i] = True
 
         # If the bracket is no longer valid, report failure (unless a function
         # tolerance is met, as detected above).
-        i = (np.sign(work.f1) == np.sign(work.f2)) & ~stop
-        work.xmin[i], work.fmin[i], work.status[i] = np.nan, np.nan, eim._ESIGNERR
+        i = (xp_sign(work.f1) == xp_sign(work.f2)) & ~stop
+        NaN = xp.asarray(xp.nan)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/optimize/_chandrupatla.py:192]: There is a `_get_nan` util which gets the NaN of the correct dtype. I don't need it here because the arrays are already typed.

File: scipy/optimize/_chandrupatla.py
Comment (mdhaber) [reply scipy/optimize/_chandrupatla.py:1]: Yup.

File: scipy/_lib/_array_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/_lib/_array_api.py:444]: Yes I read that, what I meant is, aren't all cases covered? < > and =. What is missing that needs a default blanket?

[the remark this answers is not in the document]
Comment (rkern) [reply scipy/_lib/_array_api.py:444]: NaNs.

[the remark this answers is not in the document]
Comment (tupui) [reply scipy/_lib/_array_api.py:444]: Oh, yes indeed 😅🙇

File: scipy/optimize/_chandrupatla.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/optimize/_chandrupatla.py:9]: I thought about it more carefully, and I think this is fine actually. I fixed the tests that I was having trouble with, so if tests pass, I think this is good from my side.

**Review by tupui — APPROVED** [review]
Last changes LGTM, letting the CI run and then I think it's good to merge.

**Comment by mdhaber** [comment]
Ok! Remaining failures seem unrelated - a package install conflict and slow tests, hopefully one of which is a temporary glitch.

**Comment by tupui** [comment]
Ark you have conflicts now

File: scipy/optimize/_bracket.py [hunk 5]
```diff
@@ -156,7 +156,7 @@ def _bracket_root(func, xl0, xr0=None, *, xmin=None, xmax=None, factor=None,
 
     xs = (xl0, xr0)
     temp = eim._initialize(func, xs, args)
-    func, xs, fs, args, shape, dtype = temp  # line split for PEP8
+    func, xs, fs, args, shape, dtype, _ = temp  # line split for PEP8
```
Comment (mdhaber) [inline scipy/optimize/_bracket.py:159]:
```suggestion
    func, xs, fs, args, shape, dtype, xp = temp  # line split for PEP8
```
